File: src/main/java/phoenix/world/NewEndBiomeProvider.java
package phoenix.world;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableSet;
import net.minecraft.util.SharedSeedRandom;
import net.minecraft.util.math.MathHelper;
import net.minecraft.world.World;
import net.minecraft.world.biome.Biome;
import net.minecraft.world.biome.Biomes;
import net.minecraft.world.biome.provider.BiomeProvider;
import net.minecraft.world.biome.provider.EndBiomeProviderSettings;
import net.minecraft.world.gen.IExtendedNoiseRandom;
import net.minecraft.world.gen.LazyAreaLayerContext;
import net.minecraft.world.gen.SimplexNoiseGenerator;
import net.minecraft.world.gen.area.IArea;
import net.minecraft.world.gen.area.IAreaFactory;
import net.minecraft.world.gen.area.LazyArea;
import net.minecraft.world.gen.layer.Layer;
import net.minecraft.world.gen.layer.ZoomLayer;
import net.minecraft.world.server.ServerWorld;
import phoenix.init.PhoenixBiomes;
import phoenix.init.PhoenixConfiguration;
import phoenix.world.genlayers.*;

import java.util.List;
import java.util.Set;
import java.util.function.LongFunction;

public class NewEndBiomeProvider extends BiomeProvider
{
    private final Layer genLayer;
    private final SimplexNoiseGenerator generator;
    private final SharedSeedRandom random;
    private final World world;
    private static final Set<Biome> biomes;

    static
    {
        biomes = ImmutableSet.of(Biomes.THE_END, Biomes.END_HIGHLANDS, Biomes.END_MIDLANDS, Biomes.SMALL_END_ISLANDS, Biomes.END_BARRENS, PhoenixBiomes.UNDER.get(), PhoenixBiomes.HEARTVOID.get());
    }

    public NewEndBiomeProvider(EndBiomeProviderSettings settings, World worldIn)
    {
        super(biomes);
        this.world = worldIn;
        this.genLayer = createLayer(settings.getSeed());
        this.random = new SharedSeedRandom(settings.getSeed());
        this.random.skip(17292);
        this.generator = new SimplexNoiseGenerator(this.random);
    }

      
    @Override
    public Biome getNoiseBiome(int x, int y, int z)
    {
        return this.genLayer.func_215738_a(x, z);
    }

    @Override
    public float func_222365_c(int x, int z)
    {
        int real_x = x / 2;
        int real_Z = z / 2;
        int dop_x = x % 2;
        int dop_z = z % 2;
        float result = 100.0F - MathHelper.sqrt((float) (x * x + z * z)) * 8.0F;
        result = MathHelper.clamp(result, -100.0F, 80.0F);

        for (int i = -12; i <= 12; ++i)
        {
            for (int j = -12; j <= 12; ++j)
            {
                long currect_x = real_x + i;
                long currect_z = real_Z + j;
                if (currect_x * currect_x + currect_z * currect_z > 4096L && this.generator.getValue((double) currect_x, (double) currect_z) < -0.8999999761581421D)
                {
                    float lvt_14_1_ = (MathHelper.abs((float) currect_z) * 3439.0F + MathHelper.abs((float) currect_z) * 147.0F) % 13.0F + 9.0F;
                    float lvt_15_1_ = (float) (dop_x - i * 2);
                    float lvt_16_1_ = (float) (dop_z - j * 2);
                    float lvt_17_1_ = 100.0F - MathHelper.sqrt(lvt_15_1_ * lvt_15_1_ + lvt_16_1_ * lvt_16_1_) * lvt_14_1_;
                    lvt_17_1_ = MathHelper.clamp(lvt_17_1_, -100.0F, 80.0F);
                    result = Math.max(result, lvt_17_1_);
                }
            }
        }

        return result;
    }

      
    @Override
    public List<Biome> getBiomesToSpawnIn()
    {
        return ImmutableList.copyOf(biomes);
    }

    public Layer createLayer(long seed)
    {
        IAreaFactory<LazyArea> iareafactory = getLayersApply((seedModifierIn) -> new LazyAreaLayerContext(25, seed, seedModifierIn));
        return new Layer(iareafactory);
    }

    public <T extends IArea, C extends IExtendedNoiseRandom<T>> IAreaFactory<T> getLayersApply(LongFunction<C> context)
    {
        IAreaFactory<T> phoenix_biomes = (new ParentLayer(this)).apply(context.apply(1L));
        IAreaFactory<T> vanila_biomes =  (new ParentLayer(this)).apply(context.apply(1L));
        vanila_biomes = getBiomeLayer(vanila_biomes, context);

        if(StageSaveData.get((ServerWorld) world).getStage() >= 1)
        {
            phoenix_biomes = UnderLayer.INSTANCE.apply(context.apply(200L), phoenix_biomes);
            phoenix_biomes = HeartVoidLayer.INSTANCE.apply(context.apply(200L), phoenix_biomes);
        }
        for (int i = 0; i < PhoenixConfiguration.COMMON_CONFIG.BIOME_SIZE.get(); i++)
        {
            phoenix_biomes = ZoomLayer.NORMAL.apply(context.apply(200L), phoenix_biomes);
        }

        return UnificationLayer.INSTANCE.apply(context.apply(200L), phoenix_biomes, vanila_biomes);
    }

    public <T extends IArea, C extends IExtendedNoiseRandom<T>> IAreaFactory<T> getBiomeLayer(IAreaFactory<T> parentLayer, LongFunction<C> contextFactory)
    {
        parentLayer = (new EndBiomeLayer()).apply(contextFactory.apply(200L), parentLayer);
        return parentLayer;
    }
}

File: src/main/kotlin/phoenix/world/NewEndBiomeProvider.kt
package phoenix.world

import com.google.common.collect.ImmutableBiMap
import com.google.common.collect.ImmutableList
import com.google.common.collect.ImmutableSet
import net.minecraft.util.SharedSeedRandom
import net.minecraft.util.math.MathHelper
import net.minecraft.world.biome.Biome
import net.minecraft.world.biome.Biomes
import net.minecraft.world.biome.provider.BiomeProvider
import net.minecraft.world.biome.provider.EndBiomeProviderSettings
import net.minecraft.world.gen.LazyAreaLayerContext
import net.minecraft.world.gen.SimplexNoiseGenerator
import net.minecraft.world.gen.feature.structure.Structure
import net.minecraft.world.gen.layer.Layer
import net.minecraft.world.gen.layer.ZoomLayer
import phoenix.init.PhxBiomes
import phoenix.init.PhxConfiguration
import phoenix.init.PhxFeatures
import phoenix.other.invoke
import phoenix.world.genlayers.*

class NewEndBiomeProvider(var settings: EndBiomeProviderSettings) : BiomeProvider(biomes)
{
    private lateinit var genLayer: Layer
    private val generator: SimplexNoiseGenerator
    private val random: SharedSeedRandom = SharedSeedRandom(settings.seed)

    init
    {
        this.random.skip(17292)
        this.generator = SimplexNoiseGenerator(this.random)
        initBiomeLayer()
    }

    fun initBiomeLayer()
    {
        this.genLayer = createLayer(settings.seed)
    }

    override fun getNoiseBiome(x: Int, y: Int, z: Int): Biome = genLayer.func_215738_a(x, z)

    override fun func_222365_c(x: Int, z: Int): Float
    {
        val realX = x / 2
        val realZ = z / 2
        val dopX = x % 2
        val dopZ = z % 2
        var result = 100.0f - MathHelper.sqrt((x * x + z * z).toFloat()) * 8.0f
        result = MathHelper.clamp(result, -100.0f, 80.0f)
        for (i in -12..12)
        {
            for (j in -12..12)
            {
                val currentX = realX + i.toLong()
                val currentZ = realZ + j.toLong()
                if (currentX * currentX + currentZ * currentZ > 4096L && this.generator.getValue(currentX.toDouble(), currentZ.toDouble()) < -0.8999999761581421)
                {
                    val lvt_14_1_ = (MathHelper.abs(currentZ.toFloat()) * 3439.0f + MathHelper.abs(currentZ.toFloat()) * 147.0f) % 13.0f + 9.0f
                    val lvt_15_1_ = (dopX - i * 2).toFloat()
                    val lvt_16_1_ = (dopZ - j * 2).toFloat()
                    var lvt_17_1_ = 100.0f - MathHelper.sqrt(lvt_15_1_ * lvt_15_1_ + lvt_16_1_ * lvt_16_1_) * lvt_14_1_
                    lvt_17_1_ = MathHelper.clamp(lvt_17_1_, -100.0f, 80.0f)
                    result = result.coerceAtLeast(lvt_17_1_)
                }
            }
        }
        return result
    }

    override fun getBiomesToSpawnIn() : List<Biome> = ImmutableList.copyOf(biomes)

    private fun createLayer(seed: Long): Layer
    {
        val context = { seedModifierIn: Long -> LazyAreaLayerContext(25, seed, seedModifierIn) }
        var phoenixBiomes = ParentLayer(this)(context(1L))
        val vanilaBiomes = EndBiomeLayer(context(200L), ParentLayer(this)(context(1L)))

        val stage = StageManager.stage

        if (stage >= 1)
        {
            phoenixBiomes = UnderLayer(context(200L), phoenixBiomes)
            phoenixBiomes = SmallIslandsUnderLayer(context(124L), phoenixBiomes)
            phoenixBiomes = HeartVoidLayer(context(472L), phoenixBiomes)
            phoenixBiomes = HeartVoidLayer(context(472L), phoenixBiomes)
        }

        for (i in 0..PhxConfiguration.biomeSize)
            phoenixBiomes = ZoomLayer.NORMAL(context(200L), phoenixBiomes)

        var res = UnificationLayer(context(200L), phoenixBiomes, vanilaBiomes)

        if (stage >= 2)
        {
            for (i in 0..3) res = PrimaryWetHeartVoidLayer(context(134L), res)
            for (i in 0..9) res = SecondaryWetHeartVoidLayer(context(134L), res)
        }

        return Layer(res)
    }

    companion object
    {
        private val biomes: Set<Biome> = ImmutableSet.of(Biomes.THE_END, Biomes.END_HIGHLANDS, Biomes.END_MIDLANDS, Biomes.SMALL_END_ISLANDS, Biomes.END_BARRENS, PhxBiomes.UNDER, PhxBiomes.HEARTVOID)
    }

    override fun hasStructure(structureIn: Structure<*>): Boolean = StageManager.stage >= structureToStage[structureIn] ?: 10 || super.hasStructure(structureIn)

    var structureToStage : Map<Structure<*>, Int> = ImmutableBiMap.of(PhxFeatures.REMAINS, 1)
}